hawkmoth/__init__.py

```diff
@@ -62,15 +63,54 @@ def __display_parser_diagnostics(self, errors):
                 self.logger.log(self._log_lvl[severity], toprint,
                                 location=(env.docname, self.lineno))
 
-    def __get_transform(self):
+    def __get_compat_transform(self):
         env = self.state.document.settings.env
 
         compat = self.options.get('compat', env.config.cautodoc_compat)
         if compat is None:
             return None
 
+        fmt = 'cautodoc_compat and compat options are deprecated, please use cautodoc_transformations and transform options instead.'
+        self.logger.warning(fmt, location=(env.docname, self.lineno))
+
         return lambda comment: doccompat.convert(comment, transform=compat)
 
+    def __get_transform(self):
+        env = self.state.document.settings.env
+
+        # Special case values for transform.
+        OPT_NONE = 'none'
+        OPT_DEFAULT = 'default'
+
+        # Handle deprecated compat. To be removed.
+        transform = self.__get_compat_transform()
+        if transform is not None:
+            return transform
+
+        transformations = env.config.cautodoc_transformations
+        tropt = self.options.get('transform', OPT_DEFAULT)
+
+        # Local option to skip transformation.
+        if tropt == OPT_NONE:
+            return None
+
+        if transformations is None:
+            if tropt != OPT_DEFAULT:
+                self.logger.warning('transform specified without cautodoc_transformations config.',
+                                    location=(env.docname, self.lineno))
+
+            return None
+
+        transform = transformations.get(tropt)
+
+        if transform is None:
+            if tropt != OPT_DEFAULT:
+                self.logger.warning(f'unknown transformation "{tropt}".',
+                                    location=(env.docname, self.lineno))
+            return None
+
+        return lambda comment: transform(comment, transform=tropt)
```

Comment:
Hmm, I think we discussed this in a different iteration, maybe not though. We derive the `transform` function from `tropt` and then pass `tropt` as an argument to it again. This feels redundant and makes me wonder if some extraneous look-up is being done elsewhere.

Optionally passing extra options to the transformation function may be interesting although not that useful I think (one could simply define more entries in the transformation table), but the key shouldn't matter at all. I suspect this is a remnant of how `doccompat.convert()` works.

If you want to have additional options, I would suggest something like:
```python
cautodoc_transformations = {
    'key': { 'fn': function,
             'opt': { ... } },
    ...
}
```

Reply:
Ditching the `transform` keyword option would be easy enough. Just pass it to `doccompat.convert()` as long as we have it.

What about the whole `**options` for transform functions? I'm thinking it's an easy way to pass useful information in the future. Do you suggest to remove just the one option that seems redundant, or the whole kwarg?

Reply:
Indeed, I was only referring to the tautological parameter which let's a function know that we called it. The `**options` will never be _needed_ I believe, but I like it except perhaps for how they can be defined. E.g. the suggestion I give above is too verbose in the general case (no options).

By the way, the reason I believe the options will never be a hard requirement is that one could simply (e.g.) create lambdas which captured the options we wanted and then place the lambdas (now without any extra options) on the table to no great loss.

Reply:
I was thinking of passing e.g. the Sphinx app or relevant Clang objects in the future, for more advanced stuff. Can't handle that with lambdas alone. But perhaps it's just too complicated for what Hawkmoth aims to be?

Should we just ditch the `**options` altogether? We can bring it back in the future if we need it. It's not like this is all carved in stone.

Reply:
Oh, so those options would be injected inside hawkmoth rather than through the table. In that case I'm more conflicted. On one hand if we remove it, users will have to update their transformation functions once we add it back in, if we leave it in, they may rightfully ask what it is for when they have no access to it by design.

I guess it depends on how likely you see that happening...

Reply:
I'm leaning towards dropping the `**options` out for now. If we find out it'll be useful, we can add it back. If we don't want to break backward compatibility, I think we can use inspect to figure out what we need to pass. https://docs.python.org/3/library/inspect.html#inspect.signature

Of course we could also create complex objects etc. for this, but I'd like to start simple.